File: src/main/java/club/deneb/client/features/HUDModule.java
package club.deneb.client.features;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by B_312 on 01/10/21
 */
public class HUDModule extends AbstractModule {

    public HUDModule(){
        this.x = getAnnotation().x();
        this.y = getAnnotation().y();
        this.width = getAnnotation().width();
        this.height = getAnnotation().height();
        this.name = getAnnotation().name();
        this.category = getAnnotation().category();
        this.description = getAnnotation().description();
        this.isHUD = true;
        onInit();
    }

    @Retention(RetentionPolicy.RUNTIME)
    public @interface Info {
        String name();
        int x() default 0;
        int y() default 0;
        int width() default 0;
        int height() default 0;
        Category category() default Category.HUD;
        String description() default "";
        boolean visible() default true;
    }

    private Info getAnnotation() {
        if (getClass().isAnnotationPresent(Info.class)) {
            return getClass().getAnnotation(Info.class);
        }
        throw new IllegalStateException("No Annotation on class " + this.getClass().getCanonicalName() + "!");
    }

    public void onInit(){}

    public void onDragging(int mouseX,int mouseY){}

    public void onMouseRelease(){}

}

File: src/main/java/club/deneb/client/features/HUDModule.kt
package club.deneb.client.features

/**
 * Created by B_312 on 01/10/21
 */
open class HUDModule : AbstractModule() {

    @kotlin.annotation.Retention(AnnotationRetention.RUNTIME)
    annotation class Info(
        val name: String,
        val x: Int = 0,
        val y: Int = 0,
        val width: Int = 0,
        val height: Int = 0,
        val category: Category = Category.HUD,
        val description: String = "",
        val visible: Boolean = true
    )

    private val annotation: Info
        get() {
            if (javaClass.isAnnotationPresent(Info::class.java)) {
                return javaClass.getAnnotation(Info::class.java)
            }
            throw IllegalStateException("No Annotation on class " + this.javaClass.canonicalName + "!")
        }

    open fun onDragging(mouseX: Int, mouseY: Int) {}
    open fun onMouseRelease() {}

    init {
        x = annotation.x
        y = annotation.y
        width = annotation.width
        height = annotation.height
        name = annotation.name
        this.category = annotation.category
        description = annotation.description
        isHUD = true
    }
}